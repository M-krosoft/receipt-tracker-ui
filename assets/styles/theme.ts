import { createTheme } from "@shopify/restyle";

const palette = {
  darkBlue: "#253659",
  cyan: "#03A696",
  lightCyan: "#04BF9D",
  orange: "#F28157",
  darkOrange: "#BF665E",
  white: "#fffafa",
  transparent: "#00000000",
};

const theme = createTheme({
  colors: {
    mainBackgroundGradientFirst: palette.lightCyan,
    mainBackgroundGradientSecond: palette.cyan,

    textPrimary: palette.white,
    textAccent: palette.orange,

    transparent: palette.transparent,

    buttonPrimary: palette.white,
    buttonPrimaryForeground: palette.orange,

    buttonSecondary: palette.white,
    buttonSecondaryForeground: palette.darkOrange,
  },
  spacing: {
    auto: "auto",
    xs: 4,
    s: 8,
    m: 16,
    l: 24,
    xl: 40,
    xxl: 56,
    xxxl: 64,
  },
  borderRadii: {
    xs: 2,
    s: 4,
    m: 8,
    l: 16,
    xl: 24,
  },

  textVariants: {
    header1: {
      fontWeight: "bold",
      fontSize: 34,
    },
    header2: {
      fontWeight: "bold",
      fontSize: 28,
    },
    header3: {
      fontWeight: "bold",
      fontSize: 22,
    },
    body1: {
      fontSize: 16,
    },
    body2: {
      fontSize: 14,
    },
    body3: {
      fontSize: 12,
    },
    buttonPrimary: {
      backgroundColor: "transparent",
      fontWeight: "bold",
      borderRadius: "m",
      color: "buttonPrimary",
      paddingHorizontal: "m",
      paddingVertical: "s",
    },
    buttonSecondary: {
      fontWeight: "bold",
      borderRadius: "m",
      color: "buttonPrimary",
      padding: "s",
    },
    buttonText: {
      fontWeight: "bold",
    },
    defaults: {},
  },

  cardVariants: {
    screenContainer: {
      padding: "l",
    },
    section: {
      padding: "m",
    },
    card: {
      padding: "s",
    },

    defaults: {},
  },

  buttonVariants: {
    buttonPrimary: {
      borderRadius: "m",
      backgroundColor: "buttonPrimaryForeground",
    },

    buttonSecondary: {
      borderRadius: "m",
      backgroundColor: "buttonSecondaryForeground",
    },
    buttonText: {},
    defaults: {},
  },
});

export type Theme = typeof theme;
export default theme;
