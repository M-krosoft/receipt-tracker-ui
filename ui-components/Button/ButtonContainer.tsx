import { Theme } from "@assets/styles/theme";
import {
  BoxProps,
  createRestyleComponent,
  VariantProps,
  createVariant,
} from "@shopify/restyle";
import { Box } from "@ui-components/Box/Box";

export const ButtonContainer = createRestyleComponent<
  VariantProps<Theme, "buttonVariants"> & BoxProps<Theme>,
  Theme
>([createVariant({ themeKey: "buttonVariants" })], Box);
