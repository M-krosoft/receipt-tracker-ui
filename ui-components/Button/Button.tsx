import { Theme } from "@assets/styles/theme";
import {
  useRestyle,
  spacing,
  border,
  backgroundColor,
  BackgroundColorProps,
  composeRestyleFunctions,
  TextProps,
  VariantProps,
  BoxProps,
} from "@shopify/restyle";
import { ButtonContainer } from "@ui-components/Button/ButtonContainer";
import { Text } from "@ui-components/Text/Text";
import { TouchableOpacity } from "react-native";

type RestyleProps = BoxProps<Theme> &
  BackgroundColorProps<Theme> &
  VariantProps<Theme, "buttonVariants">;

const restyleBoxFunctions = composeRestyleFunctions<Theme, RestyleProps>([
  spacing,
  border,
  backgroundColor,
]);

type Props = RestyleProps & {
  onPress: () => void;
  label: string;
  fontSize?: TextProps<Theme>["fontSize"];
};

export const Button = ({ onPress, label, fontSize, ...rest }: Props) => {
  const props = useRestyle(restyleBoxFunctions, rest);

  return (
    <TouchableOpacity onPress={onPress} accessibilityRole={"button"}>
      <ButtonContainer {...props}>
        <Text textAlign="center" fontSize={fontSize} variant={props.variant}>
          {label}
        </Text>
      </ButtonContainer>
    </TouchableOpacity>
  );
};
