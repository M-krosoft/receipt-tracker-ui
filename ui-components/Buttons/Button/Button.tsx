import { Theme } from "@assets/styles/theme";
import {
  useRestyle,
  spacing,
  border,
  backgroundColor,
  BackgroundColorProps,
  composeRestyleFunctions,
  TextProps,
  VariantProps,
  BoxProps,
} from "@shopify/restyle";
import { ButtonContainer } from "@ui-components/Buttons/Button/ButtonContainer";
import { Text } from "@ui-components/Text/Text";
import { TouchableOpacity } from "react-native";

type RestyleProps = BoxProps<Theme> &
  BackgroundColorProps<Theme> &
  VariantProps<Theme, "buttonVariants">;

const restyleBoxFunctions = composeRestyleFunctions<Theme, RestyleProps>([
  spacing,
  border,
  backgroundColor,
]);

type Props = RestyleProps & {
  onPress: () => void;
  label: string;
  fontSize?: TextProps<Theme>["fontSize"];
};

export const Button = ({ onPress, label, fontSize, ...rest }: Props) => {
  const props = useRestyle(restyleBoxFunctions, rest);

  return (
    <TouchableOpacity onPress={onPress} role={"button"} activeOpacity={0.9}>
      <ButtonContainer {...props}>
        <Text textAlign="center" fontSize={fontSize} variant={props.variant}>
          {label}
        </Text>
      </ButtonContainer>
    </TouchableOpacity>
  );
};
