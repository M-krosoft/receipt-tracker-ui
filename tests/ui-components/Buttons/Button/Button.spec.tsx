import theme from "@assets/styles/theme";
import { ThemeProvider } from "@shopify/restyle";
import { fireEvent, render } from "@testing-library/react-native";
import { Button } from "@ui-components/Buttons/Button/Button";

describe("Button", () => {
  test("should render button with correct styles", async () => {
    // When
    const renderedComponent = await render(
      <ThemeProvider theme={theme}>
        <Button
          onPress={jest.fn()}
          label="Login"
          fontSize={18}
          variant="buttonPrimary"
        />
      </ThemeProvider>
    );

    // Then
    expect(renderedComponent.toJSON()).toMatchSnapshot();
  });

  test("should call onPress when button is pressed", async () => {
    // Given
    const onPress = jest.fn();
    const renderedComponent = await render(
      <ThemeProvider theme={theme}>
        <Button onPress={onPress} label="Login" />
      </ThemeProvider>
    );

    // When
    const button = renderedComponent.getByText("Login");
    fireEvent(button, "press");

    // Then
    expect(onPress).toHaveBeenCalled();
  });
});
