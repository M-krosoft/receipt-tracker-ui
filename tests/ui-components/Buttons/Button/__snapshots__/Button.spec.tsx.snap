// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render button with correct styles 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "opacity": 1,
    }
  }
>
  <View
    style={
      [
        {},
        [
          {
            "backgroundColor": "#F28157",
            "borderRadius": 8,
          },
          [
            {},
          ],
        ],
      ]
    }
  >
    <Text
      style={
        [
          {
            "backgroundColor": "#00000000",
            "borderRadius": 8,
            "color": "#fffafa",
            "fontSize": 18,
            "fontWeight": "bold",
            "paddingHorizontal": 16,
            "paddingVertical": 8,
            "textAlign": "center",
          },
        ]
      }
    >
      Login
    </Text>
  </View>
</View>
`;
